arborio/jsonio.cpp

```diff
@@ -0,0 +1,317 @@
+    std::vector<nlohmann::json> region_vec;
+
+    nlohmann::json region;
+    for (auto it = decor.paintings().begin(); it != decor.paintings().end(); ++it) {
+        auto region_expr = to_string(it->first);
+
+        auto paintable_visitor = arb::util::overload(
+            [&](const arb::init_membrane_potential& p) { region["Vm"] = p.value; },
+            [&](const arb::axial_resistivity& p)       { region["Ra"] = p.value; },
+            [&](const arb::temperature_K& p)           { region["celsius"] = p.value - 273.15; },
+            [&](const arb::membrane_capacitance& p)    { region["cm"] = p.value; },
+            [&](const arb::init_int_concentration& p)  { region["ions"][p.ion]["internal-concentration"] = p.value; },
+            [&](const arb::init_ext_concentration& p)  { region["ions"][p.ion]["external-concentration"] = p.value; },
+            [&](const arb::init_reversal_potential& p) { region["ions"][p.ion]["reversal-potential"] = p.value; },
+            [&](const arb::mechanism_desc& p) {
+              nlohmann::json mech;
+              mech["region"] = region_expr;
+              mech["mechanism"] = p.name();
+              if (!p.values().empty()) {
+                  mech["parameters"] = p.values();
+              }
+              j["mechanisms"].push_back(mech);
+            });
+
+        std::visit(paintable_visitor, it->second);
+
+        if (region.empty()) continue;
+        auto it_next = it+1;
+        if (it == decor.paintings().end()-1 || region_expr != to_string(it_next->first)) {
+            region["region"] = region_expr;
+            region_vec.push_back(region);
+            region.clear();
+        }
+    }
+    j["local"] = region_vec;
+}
+
+void from_json(const nlohmann::json& j, decor& decor) {
+    auto j_copy = j;
+    // Global
+    if (auto globals_json = find_and_remove_json<nlohmann::json>("global", j_copy)) {
+        arb::cable_cell_parameter_set defaults;
+        try {
+            defaults = globals_json.value().get<arb::cable_cell_parameter_set>();
+        }
+        catch (std::exception& e) {
+            throw arborio::jsonio_decor_global_load_error(e.what());
+        }
+        try {
+            for (auto def: defaults.serialize()) {
+                decor.set_default(def);
+            }
+        }
+        catch (std::exception& e) {
+            throw arborio::jsonio_decor_global_set_error(e.what());
+        }
+    }
+
+    // Local
+    if (auto locals_json = find_and_remove_json<std::vector<nlohmann::json>>("local", j_copy)) {
+        for (auto l: locals_json.value()) {
+            auto region = find_and_remove_json<std::string>("region", l);
+            if (!region) {
+                throw arborio::jsonio_decor_local_missing_region();
+            }
+            std::string reg = region.value();
+
+            // if the region expression does not start with an open parenthesis,
```

Comment:
We should add documentation here and in the format proposal clarifying that 
regions are in ASCII,  _names_ must start with a letter (or number?), and that 
expressions of type `(...)` will be turned into regions via the s-exp format. Also, 
what happens if the first character is `"`?

Sorry for not bringing that up earlier.

Reply:
I attempted to do that in the [docs](https://github.com/arbor-sim/arbor/pull/1312/files?file-filters%5B%5D=.rst#diff-70cb3593cb83f3a29cd7d8710d8e01cccfb180a4cc99edbc49623db1fc19d64cR70). But I can add a comment here too. 

I am considering all region strings that don't start with an open parenthesis to be labels, which should therefore be surrounded by quotations. that should take care of all string labels that don't start with a parenthesis. So, if a user chooses to name their region `(my-region`, that's not going to work (the s-expression parser with raise an exception). I wanted to discuss this with @bcumming, I'd rather we force the double quotation in the json file than add them implicitly here, but that'll make the json file less "standard".

Reply:
That's my issue here, too, I do not think here's a good, general solution.
Maybe this could help:

- The importer checks whether a region is quoted or is an expression.
- If neither an error is raised.
- The exporter adds quotes when needed. That take care of the round trip.
- Document the need for quotes in the format spec.

Alternatively, the format could include
```
...
"region": {"kind": "s-exp", "value": "(tag 2)"}
...
"region": {"kind": "label", "value": "axon"}
...
```

Reply:
You're saying that labels should be quoted, so in the JSON file the user would write:
```
"region" : "\"soma\""
```
right? 
I don't have an issue with that except it's not very standard. Originally this JSON format was proposed as something that can be read by Arbor and Neuron. We'd be forcing an arbor-specific format in this case by requiring double quotes. 

The exporter already only prints out s-expressions. So never:
```
"region" : "\"soma\""
```
Always: 
```
"region" : "(region \"soma\")"
```
Which is potentially a problem. 

I'd like to discuss the specifics of this format with the rest of the team. All these questions are why I was uncertain about the format (and whether this PR would ultimately get merged) in the first place.

Reply:
Absolutely, I just had doubts while seeing it in code. To add a third version for consideration:

It has been a long standing point whether to allow region literals over labels in paint/place.
Maybe it is in general better to not do it and rely on the label_dict? We could add a section
to the format specification that deals with labels and allows non-portable things.

Reply:
The solution I went with for this PR may not be the best. At the moment, if the json file contains: 
```
"region" : "soma"
```
The code will check the first character of the string "soma"; notice that it's not a `(`; assume it's a label; and add quotes around it so it can be parsed correctly by arbor.

If the json file contains instead: 
```
"region" : "(my-region"
```
The code will assume that's an s-expression instead of a label and parser will fail. 

As (I think) you both suggest, maybe the best thing to do is just force the double quotes as we do everywhere else in the library. So: `"region" : "soma"` would have to be replaced by `"region" : "\"soma\""`. A neuron parser will need to remove the quotes, but that should be fine for the sake of compatibility.

Reply:
Actually my solution is too special case-y for multiple reasons. I think double-quoting is the best solution now that I've taken a closer look at our s-expression parser.